### PR TITLE
Test sequence will now 'fail fast' (#183 fix) 

### PR DESCRIPTION
**Pull request by kireledan**
This PR addresses #183 . If a step in the ```molecule test``` sequence failed that was not the ending step, the command could still return code ```0```. 

Simply put - this PR causes ```molecule test``` to abort if any sequence fails.  Also, I changed the error printing to go to stderr instead of std out in light of #180.

### CODE REVIEW

**Comment by dhutty** [comment]
👍

**Comment by rgreinho** [comment]
The patch for failing looks nice, and I do not want to be a blocker, but I'm still convinced than we should use the logging library instead of the ``file=sys.stderr`` option of the print function.

**Comment by kireledan** [comment]
I'd be fine adding that to this PR as well. Shouldn't take too long to switch it over to logging.

**Comment by retr0h** [comment]
👍

**Comment by retr0h** [comment]
@kireledan did we want to switch to logging or just go with the current patch?

**Comment by kireledan** [comment]
I think switching to logging will be the best choice. Should have the PR updated before the end of the day.

**Comment by rgreinho** [comment]
@kireledan awesome! I'll be super happy to have a look at it! thanks for doing it.

**Comment by kireledan** [comment]
@rgreinho I'm actually running into an issue with passing the unit tests though. For some reason the logging is not being picked up on the mock sys.stderr. I've tried adding a handler for sys.stderr to see if that resolved the test, but it doesn't. Is the mock sys.stderr just not able to pick it up?

**Comment by rgreinho** [comment]
@kireledan I'm not sure. Add another commit to this PR, so that I can have a look.

**Comment by kireledan** [comment]
I've pushed what I have so far. I must be missing something with the logging configuration.

File: molecule/utilities.py [hunk 1]
```diff
@@ -24,9 +24,14 @@
 import os
 import sys
 
+import logging
+
 import colorama
 import jinja2
 
+log = logging.getLogger()
+log.addHandler(logging.NullHandler())
```
Comment (rgreinho) [inline molecule/utilities.py]: Try replacing these two lines:
```
logging.basicConfig(level=logging.DEBUG)
log = logging.getLogger(__name__)
```

Comment (rgreinho) [inline molecule/utilities.py]: Maybe also renaming ``log`` with ``logger`` (nitpicking I know, but it sound clearer)

**Comment by kireledan** [comment]
Still doesn't seem to be passing the unit test.

**Comment by kireledan** [comment]
I've updated the tests to use a mock logger instead. Also I added some formatting to distinguish between warnings and errors.

**Comment by rgreinho** [comment]
The formatting should go to the logger configuration itself. for instance:
```
logging.basicConfig(format='[%(levelname).4s] %(asctime)s %(message)s')
```
Also the utilities functions ``print_error`` and ``print_warning`` do not make so much sense if we are using the logger.

**Comment by kireledan** [comment]
I have the formatting done in the separate handlers so that there is a difference in formatting for warnings and errors. The ```print_error``` and ```print_warning``` functions are still useful in that we can control the formatting / method of printing throughout the entire application with ease. Maybe change them to ```log_error``` and ```log_warning```?

**Comment by rgreinho** [comment]
The names are fine, my point was more: the abstraction provided by these 2 functions is pretty much useless. Calling the logger directly would be clearer.

**Comment by kireledan** [comment]
Yeah that's a fair point. I just didn't want to have that giant block of code that sets up the log format in each python file.

**Comment by rgreinho** [comment]
``sys.argv < 3`` does no cover all the cases. For instance, ``molecule status --porcelain`` does not display anything.

Instead of using ``sys.argv``, you may want to use the docopt args for this filtering the output of this command:
```
(Pdb) self.args
{'--debug': False,
 '--hosts': False,
 '--platforms': False,
 '--porcelain': False,
 '--providers': False,
 '<command>': 'status'}
```

Also, you must update the tab completion file if you modify the arguments of a command[0]. 
---
[0]: https://github.com/metacloud/molecule/blob/master/assets/bash_completion/molecule.bash-completion.sh#L37

**Comment by retr0h** [comment]
👍

**Pull request merged**